### PR TITLE
CI perf - prefer clang for Asan

### PR DESCRIPTION
**Pull request by daverodgman**
## Description

Tidy up use of ASAN in tests, and prefer clang instead of gcc.

The change to `test_m32_o0` improves coverage (more non-assembly paths get tested which is the point of that test).

In testing yesterday, Asan under clang looks to be around 10-20% faster than gcc, which suggests a saving of several hours of CPU time per CI run (my estimate is 145 components that use Asan, with total mean duration 1765 mins, so perhaps 2-4h of savings from using clang over gcc). HOWEVER, testing more this morning I can't reproduce so a bit uncertain about the impact now.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [x] **backport** not required
- [x] **tests** not required

### CODE REVIEW

File: tests/scripts/all.sh [hunk 1]
```diff
@@ -5227,18 +5229,20 @@ component_build_psa_alt_headers () {
...
-    make CC=gcc CFLAGS="$ASAN_CFLAGS -m32 -O0" LDFLAGS="-m32 $ASAN_CFLAGS"
+    make CC=gcc CFLAGS="$ASAN_CFLAGS -m32" LDFLAGS="-m32 $ASAN_CFLAGS"
 
-    msg "test: i386, make, gcc -O0 (ASan build)"
-    make test
+    msg "test: i386, make, gcc, no asm (ASan build)"
+    #make test
 }
-support_test_m32_o0 () {
+support_test_m32_no_asm () {
```
Comment (tom-cosgrove-arm) [inline tests/scripts/all.sh:5245]: If we're changing this, why not change it to something like `support_test_is_x86`?

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -4802,7 +4802,7 @@ component_build_aes_via_padlock () {
     scripts/config.py unset MBEDTLS_AESNI_C
     scripts/config.py set MBEDTLS_PADLOCK_C
     scripts/config.py unset MBEDTLS_AES_USE_HARDWARE_ONLY
-    make CC=$ASAN_CC CFLAGS="$ASAN_CFLAGS -m32" LDFLAGS="-m32 $ASAN_CFLAGS"
+    make CC=gcc CFLAGS="$ASAN_CFLAGS -m32" LDFLAGS="-m32 $ASAN_CFLAGS"
```
Comment (mpg) [inline tests/scripts/all.sh:4805]: Perhaps add a comment explaining why we make an exception? Also, I'm curious to know that the "known issues with clang for this target" are.

**Review by mpg — APPROVED** [review]
LGTM, thanks for finding this win!

File: tests/scripts/all.sh (hunk 2 above)
Comment (daverodgman) [reply tests/scripts/all.sh:4805]: It's this kind of link error: https://github.com/NixOS/nixpkgs/issues/201199

**Comment by daverodgman** [comment]
Actually, testing it again locally this morning I'm unsure that clang is in fact faster. I'm inclined to merge this anyway, and revisit next week to see what the impact is (I have a script that can extract average component times from PR jobs) - obviously this is now very easy to change back to gcc if that turns out to be better.

File: tests/scripts/all.sh (hunk 1 above)
Comment (daverodgman) [reply tests/scripts/all.sh:5245]: It has to have a name corresponding to the test.

**Review by mpg — APPROVED** [review]

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by daverodgman** [comment]
OpenCI results seem more supportive: comparing this to a recent PR run without this change:

https://mbedtls.trustedfirmware.org/blue/rest/organizations/jenkins/pipelines/mbed-tls-pr-head/branches/PR-8623-head/runs/5/nodes/431/log/?start=0 vs https://mbedtls.trustedfirmware.org/blue/rest/organizations/jenkins/pipelines/mbed-tls-pr-head/branches/PR-8635-head/runs/6/nodes/431/log/?start=0 there is a big difference in test execution times (quick look at a few of the slow tests):

[2023-12-13T18:04:30.962Z]  54/120 Test  #54: gcm.aes128_en-suite ........................   Passed    5.01 sec
[2023-12-13T18:05:47.114Z] 103/120 Test #103: psa_crypto_pake-suite ......................   Passed   16.83 sec
[2023-12-13T18:07:39.148Z] 110/120 Test #110: psa_crypto_storage_format.v0-suite .........   Passed   99.94 sec
[2023-12-13T18:07:39.148Z] 115/120 Test #115: ssl-suite ..................................   Passed    7.56 sec
[2023-12-13T18:07:45.719Z] 116/120 Test #116: ssl_decrypt.misc-suite .....................   Passed    6.38 sec

vs

[2023-12-15T11:14:36.447Z]  54/120 Test  #54: gcm.aes128_en-suite ........................   Passed    3.69 sec
[2023-12-15T11:15:15.219Z] 103/120 Test #103: psa_crypto_pake-suite ......................   Passed    1.48 sec
[2023-12-15T11:15:33.852Z] 110/120 Test #110: psa_crypto_storage_format.v0-suite .........   Passed   15.19 sec
[2023-12-15T11:15:40.793Z] 115/120 Test #115: ssl-suite ..................................   Passed    5.15 sec
[2023-12-15T11:15:46.072Z] 116/120 Test #116: ssl_decrypt.misc-suite .....................   Passed    5.78 sec

The CI is using clang 3.8 and gcc 5.4 here, so maybe we only see the difference with old compiler versions.

**Comment by daverodgman** [comment]
Results for a PR with this enabled show total CPU time of 34h, vs. 44+/-2.7. Total CPU time is a fairly consistent metric so looks like a saving of around 10h.

**Pull request merged**